var/spack/repos/builtin/packages/random123/package.py

```diff
@@ -22,6 +22,7 @@ class Random123(Package):
     patch('ibmxl.patch', when='@1.09')
     patch('arm-gcc.patch', when='@1.09')
     patch('v1132-xl161.patch', when='@1.13.2')
+    conflicts('%nvhpc')
```

Comment:
This seems wrong. As far as I understand from the recipe this package is a C++ template library, so the conflict shouldn't be here but rather in clients of this library.

Reply:
sorry if this is not the correct fix, but how should we reflect this issue? Random123 has this file:
https://github.com/DEShawResearch/random123/blob/main/include/Random123/philox.h
And the way random123 is used is that a package X has it as dependency, and might use this (or other headers) => in that case it will conflict with 'nvhpc', so it's not that Random123 itself won't 'spack install' because there is no compilation involved, _but_ if package X depends on Random123, then X will conflict with 'nvhpc', if Random123 is set to conflict with nvhpc, no such error will happen => 
If we set Random123 as conflicting with nvhpc, every X which depends on Random123 will get a nice error message if you try building with nvhpc, currently, you don't get nice error message, but a long confusing trace of compilation errors. 
The package in question is NuT (but there might be more similar situations). Now do you suggest I set that 'NuT' package as conflicting with nvhpc instead of Random123 instead?

Reply:
No need to apologize, I just meant to start a discussion with the comment above :slightly_smiling_face:  I also have to say that this is partially a lack in our model of software for C++ templated library, in the sense that we don't have an easy way to _inject_ a conflict from a dependency to a dependent.

> if Random123 is set to conflict with nvhpc, no such error will happen

This is not true technically as the only node that conflicts with `%nvhpc` is that of the templated library, so you might have:
```
X%nvhpc ^random123%gcc
```
and that will be legitimate to Spack.

>  Now do you suggest I set that 'NuT' package as conflicting with nvhpc instead of Random123 instead?

If there's only one package depending on this one lets move the conflict to `NuT`. When they start to be more than a handful we'll need some support in the DSL (maybe add a new directive?).

Reply:
Thanks!
done: https://github.com/spack/spack/pull/25023